openraft/src/core/admin.rs

```diff
@@ -73,15 +73,48 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
 }
 
 impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> LeaderState<'a, D, R, N, S> {
+    #[tracing::instrument(level = "debug", skip(self))]
+    async fn add_learner_into_membership(&mut self, target: &NodeId) {
+        tracing::debug!(
+            "before add_learner_into_membership target node {} into learner {:?}",
+            target,
+            self.nodes.keys()
+        );
+
+        let membership = &self.core.effective_membership.membership;
+        if membership.is_member(target) {
+            tracing::debug!("target node {} is already a member,cannot add as learner", target);
+            return;
+        }
+
+        let curr = &mut self.core.effective_membership.membership;
+
+        if curr.is_learner(target) {
+            tracing::debug!("target node {} is already a learner", target);
+
+            return;
+        }
+        curr.add_learner(target);
+
+        // append_membership_log after learner has added into nodes
+        let new_config = self.core.effective_membership.membership.clone();
+
+        tracing::debug!(?new_config, "new_config");
+
+        let _ = self.append_membership_log(new_config, None).await;
```

Comment:
`self.core.effective_membership` should only be updated by `append_payload_to_log()` which is called by `append_membership_log()`.

Before the log entry is written `effective_membership` should not be updated.